File: junit/se/fredsberg/game/container/PoolTest.java
package se.fredsberg.game.container;

import static org.junit.Assert.*;

import java.util.List;

import org.junit.Test;

import se.fredsberg.game.random.ReverseScrambler;

public class PoolTest {

    @Test
    public void getCount() {
        assertEquals(0, new Pool<Integer>().getCount());
    }

    @Test
    public void isEmpty() {
        Pool<Integer> pool = new Pool<Integer>();
        assertTrue(pool.isEmpty());
        pool.add(6);
        assertFalse(pool.isEmpty());
    }

    @Test(expected = IllegalArgumentException.class)
    public void add_nullNotAcceptable() {
        Pool<Integer> pool = new Pool<Integer>();
        pool.add(null);
    }

    @Test
    public void add() {
        Pool<Integer> pool = new Pool<Integer>();
        pool.add(6);
        assertEquals(1, pool.getCount());
    }

    @Test
    public void add_checkFifoOrder() {
        Pool<Integer> pool = new Pool<Integer>();
        pool.add(6);
        pool.add(9);
        assertPoolOrder(pool, 6, 9);
    }

    @Test(expected = EmptyException.class)
    public void pickOne_emptyShouldResultInException() {
        Pool<Integer> pool = new Pool<Integer>();
        pool.pickOne();
    }

    @Test
    public void pickOne() {
        Pool<Integer> pool = new Pool<Integer>();
        pool.add(6);
        assertEquals((Integer) 6, pool.pickOne());
    }

    @Test
    public void pickAll_shouldBeEmptyAfterwards() {
        Pool<Integer> pool = new Pool<Integer>();
        pool.add(6);
        List<Integer> all = pool.pickAll();
        assertEquals(1, all.size());
        assertEquals((Integer) 6, all.iterator().next());
        assertTrue(pool.isEmpty());
    }

    @Test
    public void shake_inRevereseOrderWhenReverseScramblerIsUsed() {
        Pool<Integer> pool = new Pool<Integer>();
        pool.add(1);
        pool.add(2);
        pool.shake(new ReverseScrambler());
        assertPoolOrder(pool, 2, 1);
    }

    private void assertPoolOrder(Pool<Integer> pool, int... index) {
        for (int i : index) {
            assertSame(i, pool.pickOne());
        }
    }
}

File: jstests/PoolTest.js
var assert = require('assert');
var Pool = require('../jssrc/Pool.js');
var ReverseScrambler = require('./ReverseScrambler.js');
module.exports = {
    GetCount: function() {
        assert.equal(0, new Pool().getCount());
    },
    IsEmpty: function() {
        assert.equal(true, new Pool().isEmpty());
    },
    Add: function() {
        var pool = new Pool();
        pool.add(6);
        assert.equal(false, pool.isEmpty());
        assert.equal(1, pool.getCount());
    },
    PickAll_shouldBeEmptyAfter: function() {
        var pool = new Pool();
        pool.add(6);
        var all = pool.pickAll();
        assert.equal(1, all.length);
        assert.equal(6, all[0]);
        assert.equal(true, pool.isEmpty());
    },
    PickOne: function() {
        var pool = new Pool();
        pool.add(6);
        assert.equal(6, pool.pickOne());
    },
    Add_checkFifoOrder: function() {
        var pool = new Pool();
        pool.add(6);
        pool.add(9);
        assert.equal(6, pool.pickOne());
        assert.equal(9, pool.pickOne());
    },
    Pickone_undefinedIfEmpty: function() {
        var pool = new Pool();
        assert.equal(undefined, pool.pickOne());
    },
    Shake_inRevereseOrderWhenReverseScramblerIsUsed: function() {
        var pool = new Pool();
        pool.add(1);
        pool.add(2);
        pool.shake(new ReverseScrambler());
        assert.equal(2, pool.pickOne());
        assert.equal(1, pool.pickOne());
    }
};
